### PR TITLE
Add in options[:scope] and receiving scope in attribute blocks

### PR DESCRIPTION
**Pull request by bbuck**
This feature is an alternative implementation to #150 that also should provide a solution for #139 and #145 and, potentially, #148.

Why I chose a different path: #150 opts to use a class-level `attr_accessor` for setting the scope, this approach feels open to potential race conditions or overlap as the scope may persist to other serializer instances or having more instances of one kind of serializer may unintentionally remove a scope -- basically the class-level approach feels open to unnecessary errors.

This solution instead opts for a local instance variable to the serializer instance and pass it upward to the class-level methods as necessary like it does for `@resource`.

**NOTE** There are some failing performance specs but they appear to be failing (at least for me) without my change. I wasn't able to notice any significant amount of decrease when my changes were present by if you do see any speed-ups please point them out!

### CODE REVIEW

**Comment by jshow** [comment]
thank you @bbuck - I like this approach - I will review, good to get some eyes and thoughts on this

**Comment by shishirmk** [comment]
@bbuck Thank you for working on this. I like this approach too. Do you see anything we could do to support #148 as well?

**Comment by jshow** [comment]
@bbuck, to echo @christophersansone  comments in #148 , it may make sense to put the custom user options inside a standard subhash - `user_options` or something - to remove the chance of namespace collisions with keys used by fast_jsonapi

```
UserSerializer.new(model, {user_options: {scope: current_user, use_token: true})
```

**Comment by bbuck** [comment]
@jshow Technically speaking that's what `scope` is. The point of the `:scope` key is to provide a context for the JSON to serialize in. In my example it's `current_user` but if your example:

```ruby
UserSerializer.new(model, {user_options: {scope: current_user, use_token: true})
```

You're scope should probably look more like:

```ruby
UserSerializer.new(model, {scope: {user: current_user, use_token: true})

# use
attribute :token do |user, scope|
  if scope[:use_token]
    user.get_token
  end
end
```

I'm not attached to the name `:scope` but it has clear meaning for it's purpose which is why I chose it. I don't see how renaming or nesting it will really change or fix anything unless scope becomes a portion of the JSON:API spec at some point (which seems unlikely).

**Comment by bbuck** [comment]
@shishirmk I'm not sure if you've had a chance to review my comment above. The `scope` value is relatively open, and best practice is probably to always assign it a hash/struct-like object (so that future additions don't require rewriting the serializer). This can cover options, context variables, etc... required when serializing.

**Comment by jdongelmans** [comment]
Any updates? I agree with @bbuck that every related issue can be solved using the ```scope``` parameter 👍

**Comment by bbuck** [comment]
1) Totally agree, I was shooting for simplicity in initial examples but have since begun using a hash in the project in which I needed these changes too. Is the best approach for enforcing a hash to throw an error if it's not a hash?

2) I think so, but if the consensus is for it to change that's fine. What I see is exactly as you defined it, "the context in which code is being executed" as the scope contains data that defines the context in which the resource is being serialized. But I'm not going to hold up a feature for a name, if a better name is agreed on then it's easy to change.

**Comment by christophersansone** [comment]
@bbuck Thanks!  To answer your questions:

1.  Yeah just throw an `ArgumentError` with a detailed message `unless params.is_a?(Hash)`.

2.  Agreed that my definition was lacking, and agreed that it's not that big of a deal either way, but the more precise we can make the term, the more easily it can be understood without a lot of explanation.  The difference between `scope` and `params` is subtle but I believe it's there.  I would describe the difference as this: `scope` contains the variables and methods _built into the context_, while `params` contain the variables and methods _passed into the context_.  The term `params` makes it clear that these values originated externally and were passed in, which is what we are doing here.  In the future, we intend to allow developers to define methods within the serializer class to calculate attribute values and such, and these would fall more into the category of scope.  [Suggested reading.](https://blog.codeship.com/a-deep-dive-into-ruby-scopes/) .

@shishirmk what do you think?  Anyone else have an opinion?

**Comment by kyreeves** [comment]
First off, I'm loving this gem and thank you for all the work you all have done!

This PR would be very helpful, but I was wondering if there would be a way to use the scope (or params) to conditionally add attributes. You can see what I’m talking about [here in the AMS docs](https://github.com/rails-api/active_model_serializers/blob/v0.10.6/docs/general/serializers.md#attribute). I’m able to do something like this with #150 because I can access the scope on the class level, but after some trial and error, I’m having a hard time trying to implement it with this PR's approach.

**Comment by christophersansone** [comment]
@kyreeves Thanks for the suggestion!  I'd say that is out of scope for this particular issue.  Here we are discussing enhancing _how_ the attribute values are determined, not determining _if_ the attributes should be included at all.  We can support a syntax similar to AMS to include an `if` proc, but it's outside the scope here.  Feel free to create a new issue!

**Comment by kyreeves** [comment]
@christophersansone understood! Will do.

**Comment by shishirmk** [comment]
@bbuck @christophersansone I like this approach over other approaches. I think we should allow for it to be generic and use the word `params` as @christophersansone suggested.

@bbuck once you make that change happy to merge it in.

**Comment by bbuck** [comment]
@shishirmk Changes are in place!

**Comment by mhluska** [comment]
Was there any reason for making params default to `nil` instead of `{}`? I'm finding I have to do `params && params[:foo]` in all my serializers now because it's not certain whether the caller will explicitly pass in a hash or not.

**Comment by bbuck** [comment]
With the strict Hash enforcement now in place that probably makes more sense than `nil`. Does anyone else have an opinion.

**Comment by bbuck** [comment]
cc @mhluska forgot to tag you

**Comment by christophersansone** [comment]
@mhluska Good call!  Yes, now that `params` is a bona fide property, the developer should always be able to expect a hash.  It will feel similar to Rails controllers: `params` is always available, even if no request parameters were specified.

@bbuck Thanks for your work on this, you rock.  Would you mind making this adjustment?

The only other thing I think would be helpful is making `params` a public method.  It feels like a part of the public API at this point, and soon we will be adding the ability to define custom methods in the serializer for calculating attributes, relationships, etc.  @bbuck What do you think about adding an `attr_reader :params` to `ObjectSerializer`?

**Comment by bbuck** [comment]
@christinaarmstrong With attributes being serialized at the class level, what benefit does an instance method to access params provide?

**Comment by christophersansone** [comment]
@bbuck Probably nothing yet... I'm thinking more along the lines of #71 and being able to write custom methods within the serializer.  I'm not sure how useful it is outside of that, but in general it just seems like _a serializer has params_.

Take it a step further, and we can execute attribute blocks within the context of the serializer, giving us access to the `params` method without having to pass it in.  In other words, we can go from:

```ruby
attribute :viewed do |movie, params|
  params[:viewed_movies].include?(movie)
end
```

to:

```ruby
attribute :viewed do |movie|
  params[:viewed_movies].include?(movie)
end
```

Either way I'd say *mission accomplished* for this ticket, and we can choose to defer this until we have more reason to do so.

**Comment by bbuck** [comment]
@christophersansone I'm not against any of those suggestions. My hesitations was with the current implementation not really requiring it. I think that's a great end goal to aspire too in simplifying the API for users (not having to know to provide a block requiring two arguments vs. one). I definitely liked that aspect of the #150, now #171.

**Comment by christophersansone** [comment]
If a serializer instantiates other serializers, I would expect that ALL `options` would be passed down.  So yes, `params` should, and so should everything else... right?

**Comment by mhluska** [comment]
I'm still having issues with the default params hash in some cases even after the latest changes here. But there's a possibility that it's caused by some cherry-picked PRs in my branch (I'm using the nested includes PR as well). Just thought I'd point it out.

**Comment by bbuck** [comment]
@mhluska If you can be more specific I can potentially review and add test cases to verify that those concerns will be honored.

**Comment by mhluska** [comment]
@bbuck in my case I have a `User` model with a `has_one :profile` in the serializer. In my user_controller show action, I do this:

```
render json: CurrentUserUserSerializer.new(
  User.includes(
    :profile,
    ...
    ).find(params[:id]),
    include: [:profile],
    params: { scope: current_user }
```

But then in the profile serializer, I try to rely on `params[:scope]` but it doesn't exist. My first thought is this is related to the `include` option but I haven't had the time to debug it.

File: lib/fast_jsonapi/object_serializer.rb [hunk 1]
```diff
@@ -72,6 +72,8 @@ def process_options(options)
       @known_included_objects = {}
       @meta = options[:meta]
       @links = options[:links]
+      @params = options[:params] || {}
+      raise ArgumentError.new("`params` option passed to serializer must be a hash") unless @params.is_a?(Hash)
```
Comment (zinosama) [inline lib/fast_jsonapi/object_serializer.rb]: A bit late to the party, but i'm curious why params has to be a hash?

Comment (bbuck) [reply lib/fast_jsonapi/object_serializer.rb]: @zinosama Please review discussion starting here: https://github.com/Netflix/fast_jsonapi/pull/153#issuecomment-382211282

Comment (zinosama) [reply lib/fast_jsonapi/object_serializer.rb]: Awesome. Reasonable. Thanks for the quick response 😄

Comment (bbuck) [reply lib/fast_jsonapi/object_serializer.rb]: Yea no worries, sorry I did still link you to the wrong starter conversation but it seems you found the full convo!

Comment (zinosama) [reply lib/fast_jsonapi/object_serializer.rb]: I opened https://github.com/Netflix/fast_jsonapi/pull/200 and meant to @ you but didn't find you in the @-list. It'd be reassuring if you could take a quick look and make sure i didn't miss anything?

**Comment by theDisco** [comment]
Hey, thanks for implementing this very needed feature. I am waiting for it and currently installing the gem from the dev branch. Do you know when it is planned to release it?

**Comment by arefaslani** [comment]
@shishirmk I really need this... When do you release this feature?

**Comment by shishirmk** [comment]
@theDisco @arefaslani plan to release 1.2.0 this weekend

**Comment by albertusnicholas** [comment]
Thank you so much we are currently use FastJson api for our project but we stuck with this feature that we need hope can use. Pleaseeeeeeeee :100:

**Comment by dwiliandy** [comment]
Thank you so much, we need this feature for our project...

**Comment by leonmojica** [comment]
Thank you very much for this fine piece of software, it sure makes my life a lot easier, any ideas when this will be released?

**Comment by arefaslani** [comment]
> @theDisco @arefaslani plan to release 1.2.0 this weekend

**Pull request merged**